design/one-pager-package-signature-validation.md

```diff
@@ -0,0 +1,320 @@
+
+- https://github.com/sigstore/cosign/tree/1e055235dc5bdab5f9d446ad5fd31dbc72225312/copasetic#cosignverifyref-pubkey-cosignsignedpayload
+
+- https://github.com/sigstore/cosign/blob/1e055235dc5bdab5f9d446ad5fd31dbc72225312/copasetic/main.go#L190-L201
+
+- https://github.com/sigstore/cosign/blob/8ffcd1228c463e1ad26ccce68ae16deeca2960b4/pkg/cosign/kubernetes/webhook/validation.go#L60-L85
+
+Also in tekton-chains:
+- https://github.com/tektoncd/chains/blob/main/docs/tutorials/signed-provenance-tutorial.md
+
+- https://github.com/tektoncd/chains/blob/main/docs/signing.md
+
+#### Wiring Up Package Fetcher:
+As it turns out there is a great example PR for how we can pass the public key
+into the fetcher. The PR for adding an alternate CA Bundle is very similar:
+https://github.com/crossplane/crossplane/pull/2525/files
+
+Adding a new optional fetcher parameter using the builder pattern for the
+fetcherOptions struct seems pretty straightforward:
+```go
+// FetcherOpt can be used to add optional parameters to NewK8sFetcher
+type FetcherOpt func(k *K8sFetcher) error
+```
+[Link]: https://github.com/crossplane/crossplane/blob/b01b17353198a8de28664cf1eec601aaaf2fd95a/internal/xpkg/fetch.go#L63
+
+```go
+// WithCustomCA is a FetcherOpt that can be used to add a custom CA bundle to a K8sFetcher
+func WithCustomCA(rootCAs *x509.CertPool) FetcherOpt {
+	return func(k *K8sFetcher) error {
+		t, ok := k.transport.(*http.Transport)
+		if !ok {
+			return errors.New("Fetcher transport is not an HTTP transport")
+		}
+
+		t.TLSClientConfig = &tls.Config{RootCAs: rootCAs, MinVersion: tls.VersionTLS12}
+		return nil
+	}
+}
+```
+[Link]: https://github.com/crossplane/crossplane/blob/b01b17353198a8de28664cf1eec601aaaf2fd95a/internal/xpkg/fetch.go#L89-L99
+
+These can be wired up using the plumbing for options passing to controller @negz
+introduced in the following commit, which feeds it from Kong startCommand:
+```go
+if c.CABundlePath != "" {
+	rootCAs, err := xpkg.ParseCertificatesFromPath(c.CABundlePath)
+	if err != nil {
+		return errors.Wrap(err, "Cannot parse CA bundle")
+	}
+	po.FetcherOptions = []xpkg.FetcherOpt{xpkg.WithCustomCA(rootCAs)}
+}
+```
+[Link]: https://github.com/crossplane/crossplane/blob/fc4619685d1564fafee70a8456fcc124133570e0/cmd/crossplane/core/core.go#L111-L11
+
+We should probably add a test case. Seems like there is a fetch failure test
+here:
+```go
+"ErrBadFetch": {
+	reason: "Should return an error if we fail to fetch package image.",
+	args: args{
+		f: &fake.MockFetcher{
+			MockHead: fake.NewMockHeadFn(nil, errBoom),
+		},
+		pkg: &v1.Provider{
+			Spec: v1.ProviderSpec{
+				PackageSpec: v1.PackageSpec{
+					Package: "test/test:test",
+				},
+			},
+		},
+	},
+	want: want{
+		err: errors.Wrap(errBoom, errFetchPackage),
+	},
+},
+```
+[Link]: https://github.com/crossplane/crossplane/blob/master/internal/controller/pkg/manager/revisioner_test.go#L113-L131
+
+## Alternatives Considered
```

Comment:
It might also be possible to utilise Flux's SourceController in a similar way to what Pulumi is doing in Kubernetes Operator.

See:
- https://github.com/pulumi/pulumi-kubernetes-operator/pull/324
- https://www.pulumi.com/blog/pulumi-kubernetes-new-2022/

Reply:
It might also be possible to utilise Flux's source-controller / OCIRepository in a similar way to what Pulumi are doing in their Kubernetes Operator.

See:

- [Flux source integration](pulumi/pulumi-kubernetes-operator#324)
- https://www.pulumi.com/blog/pulumi-kubernetes-new-2022/

Reply:
> "Could the stated goals be achieved with an external system such as OPA Gatekeeper or Kyverno? These external systems, where used, would be validating the OCI images in the deployments for providers."

 Unfortunately, the image pull operations are done within the crossplane controller, and the image signature can't be verified until the image is pulled. So this all has to be evaluated after "admission". The same applies for config packages and provider packages.

Reply:
Could patterns not be defined for OPA Gatekeeper or Kyverno to validate the image references in the Crossplane resources?

Reply:
@stevehipwell yes I imagine that we could have them watch for the package install CRDs and deny them if they don't contain valid signature information, but how would we validate the signatures are correct? The evaluation of the signature can only be done once the image has been downloaded. That pull of the image is all internal to the crossplane controllers.